### PR TITLE
[mir-opt] asking `?`s in a more optimized fashion

### PR DESCRIPTION
**Pull request by Centril**
This PR works towards https://github.com/rust-lang/rust/issues/66234 by providing two optimization passes meant to run in sequence:

- `SimplifyArmIdentity` which transforms something like:
  ```rust
  _LOCAL_TMP = ((_LOCAL_1 as Variant ).FIELD: TY );
  ((_LOCAL_0 as Variant).FIELD: TY) = move _LOCAL_TMP;
  discriminant(_LOCAL_0) = VAR_IDX;
  ```
 
  into:
 
  ```rust
  _LOCAL_0 = move _LOCAL_1
  ```

- `SimplifyBranchSame` which transforms `SwitchInt`s to identical basic blocks into a `goto` to the first reachable target.

Together, these are meant to simplify the following into just `res`:
```rust
match res {
    Ok(x) => Ok(x),
    Err(x) => Err(x),
}
```

It should be noted however that the desugaring of `?` includes a function call and so the first pass in this PR relies on inlining to substitute that function call for identity on `x`. Inlining requires `mir-opt-level=2` so this might not have any effect in perf-bot but let's find out.

r? @oli-obk -- This is WIP, but I'd appreciate feedback. :)

### CODE REVIEW

File: src/test/mir-opt/simplify_try.rs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+//         }
+//     }
+//     scope 4 {
+//         scope 5 {
+//         }
+//     }
+//     scope 6 {
+//     }
+//     bb0: {
+//         _2 = discriminant(_1);
```
Comment (Centril) [inline src/test/mir-opt/simplify_try.rs]: This is a tad unfortunate that `_2` isn't optimized out.

File: src/librustc_mir/transform/simplify_try.rs [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+                    && bb_l.statements.iter().eq_by(&bb_r.statements, |x, y| x.kind == y.kind)
+                });
+
+            if all_successors_equivalent {
+                // Replace `SwitchInt(..) -> [bb_first, ..];` with a `goto -> bb_first;`.
+                bbs[bb_idx].terminator_mut().kind = TerminatorKind::Goto { target: bb_first };
+            }
+        }
+
+        // We may have dead blocks now, so remvoe those.
```
Comment (Centril) [inline src/librustc_mir/transform/simplify_try.rs]:
```suggestion
        // We may have dead blocks now, so remove those.
```

File: src/librustc_mir/transform/mod.rs [hunk 3]
```diff
@@ -275,6 +276,9 @@ fn run_optimization_passes<'tcx>(
         &copy_prop::CopyPropagation,
         &simplify_branches::SimplifyBranches::new("after-copy-prop"),
         &remove_noop_landing_pads::RemoveNoopLandingPads,
+        &simplify::SimplifyCfg::new("after-remove-noo-landing-pads"),
```
Comment (Centril) [inline src/librustc_mir/transform/mod.rs]:
```suggestion
        &simplify::SimplifyCfg::new("after-remove-noop-landing-pads"),
```

File: src/librustc_mir/transform/simplify_try.rs [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        simplify::remove_dead_blocks(body);
+    }
+}
+
+/*
+
+KEEPSAKE: REMOVE IF NOT NECESSARY!
```
Comment (Centril) [inline src/librustc_mir/transform/simplify_try.rs]: If we are OK with the `PartialEq`s then this is garbage that can be removed.

File: src/librustc_mir/transform/simplify_try.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_mir/transform/simplify_try.rs]: Not necessarily. This should only be run if anything was changed

File: src/librustc_mir/transform/simplify_try.rs [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+            // We want to `goto -> bb_first`.
+            let bb_first = iter_bbs_reachable
+                .peek()
+                .map(|(idx, _)| *idx)
+                .unwrap_or(targets[0]);
+
+            // All successor basic blocks should have the exact same form.
+            let all_successors_equivalent = iter_bbs_reachable
+                .map(|(_, bb)| bb)
+                .tuple_windows()
```
Comment (oli-obk) [inline src/librustc_mir/transform/simplify_try.rs]: If this is the only use case for itertools that seems a bit overkill to include

File: src/test/mir-opt/simplify_try.rs (hunk 1 above)
Comment (oli-obk) [reply src/test/mir-opt/simplify_try.rs]: I thought we had some sort of DCE. If not, open an issue so we create it

File: src/librustc_mir/transform/simplify_try.rs (hunk 5 above)
Comment (Centril) [reply src/librustc_mir/transform/simplify_try.rs]: It's already included in the lockfile and besides, if we never use it we'll just end up with boilerplate instead.

File: src/librustc_mir/transform/simplify_try.rs (hunk 2 above)
Comment (Centril) [reply src/librustc_mir/transform/simplify_try.rs]: Good point; will fix.

File: src/test/mir-opt/simplify_try.rs (hunk 1 above)
Comment (wesleywiser) [reply src/test/mir-opt/simplify_try.rs]: I think this will be easy to fix. I'll open a PR tonight with that.

Comment (oli-obk) [reply src/test/mir-opt/simplify_try.rs]: There are some ppl at rustfest working on it

**Comment by Centril** [comment]
Adjusted the existing `match.rs` codegen test and introduced a new one which demonstrates that the MIR optimizations introduced produce better LLVM IR as compared to [before](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=e703da8ad0506a99ed19dc566081869f).

---------------------

Me and Oliver have been discussing some generalizations to these optimizations (for follow ups). In particular, we can generalize the second one into two passes:

1. `DedupBlocks` -- This one looks for identical basic blocks, collects a substitution map, and then substitutes all references to `BasicBlock` for the key in the map. If we want, this could be run to a fixed point, but we could start it off as a one-shot pass.

2. `SwitchToSamePlace` -- This does (could be split up but maybe not depending on complexity):
    a. Filters BBs that are unreachable
    b. Merges identical targets
    c. If one target is left ==> goto
    d. If no targets are left ==> unreachable

   -- a. & d. sorta want to be run to a fixed point but we should probably start off as one-shot instead.

Finally, the `SimplifyArmIdentity` could be generalized to also recognize e.g. `Var(a, b, c) => Var(a, b, c)` and not just `Var(a) => Var(a)`.

**Comment by Centril** [comment]
Actually, looking at `CfgSimplifier::simplify_branch` it looks like `SwitchToSamePlace` is a generalization of that logic. `simplify_branch` requires *all* targets to end up in the same place whereas the generalization would first do a filter and simplify e.g. `-> [bb2, bb3, otherwise: bb2]` to `-> [bb2, otherwise: bb3]`.

File: src/test/mir-opt/simplify_try.rs (hunk 1 above)
Comment (Centril) [reply src/test/mir-opt/simplify_try.rs]: Does that refer to https://github.com/rust-lang/rust/pull/66329 ? -- That won't help in this case since this is an unused, but reachable, local.

Comment (oli-obk) [reply src/test/mir-opt/simplify_try.rs]: No, different optimization, not sure on what repo it's being developed

**Comment by Centril** [comment]
This should be ready now :)

File: src/librustc_mir/transform/simplify_try.rs [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            let targets = match &bbs[bb_idx].terminator().kind {
+                TerminatorKind::SwitchInt { targets, .. } => targets,
+                _ => continue,
+            };
+
+            // Reaching `unreachable` is UB so assume it doesn't happen.
+            let mut iter_bbs_reachable = targets
+                .iter()
+                .map(|idx| (*idx, &bbs[*idx]))
+                .filter(|(_, bb)| bb.terminator().kind != TerminatorKind::Unreachable)
```
Comment (Centril) [inline src/librustc_mir/transform/simplify_try.rs]: @rkruppe So we would need to check that there are no `InlineAsm` statements here wrt. the point you made on the other PR?

File: src/test/codegen/try_identity.rs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+// compile-flags: -C no-prepopulate-passes -Z mir-opt-level=2
```
Comment (oli-obk) [inline src/test/codegen/try_identity.rs]: Does this also work without -Z flags? Is this directly exposed to users via `--release`? Or should that be done in a separate PR?

File: src/librustc_mir/transform/simplify_try.rs (hunk 6 above)
Comment (rkruppe) [reply src/librustc_mir/transform/simplify_try.rs]: Yes. Though since there's at least two passes that need this information, I would recommend again to implementing that in a way that can be shared between the passes.

BTW, I'm starting to think it might be more elegant to treat this as a canonicalization rather than a check. That is, instead of repeatedly checking whether there's any inline asm, have a rewrite that removes non-divering statements leading up to a `Unreachable`. Then this and other passes can run the canonicalization (which can be made very cheap if nothing changed) and then just check for `bb.statements().is_empty() && bb.terminator().kind == Unreachable`.

File: src/test/codegen/try_identity.rs (hunk 7 above)
Comment (Centril) [reply src/test/codegen/try_identity.rs]: No idea. =) `-Z mir-opt-level=2` needed to trigger inlining, which this test needs. cc @wesleywiser 

We could maybe special case inlining for some functions via attributes?

Comment (wesleywiser) [reply src/test/codegen/try_identity.rs]: I'm not sure the inline pass is stable enough to offer even via an opt-in attribute.

Comment (Centril) [reply src/test/codegen/try_identity.rs]: @wesleywiser just to clarify, I meant `#[rustc_inline_me]` just for use in the standard library or something.

Comment (wesleywiser) [reply src/test/codegen/try_identity.rs]: @Centril Yeah, that's what I thought you meant. I just think the inliner has a lot of sharp edges and I'd rather not give people tools *this* sharp.

Comment (Centril) [reply src/test/codegen/try_identity.rs]: @wesleywiser Fair enough; Is there a dumber "definitely correct" version of the inliner we could use for "just this function" which doesn't try to inline recursively and whatnot?

Comment (wesleywiser) [reply src/test/codegen/try_identity.rs]: Possibly yeah :)

File: src/librustc_mir/transform/simplify_try.rs [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        for bb_idx in bbs.indices() {
+            let targets = match &bbs[bb_idx].terminator().kind {
+                TerminatorKind::SwitchInt { targets, .. } => targets,
+                _ => continue,
+            };
+
+            let mut iter_bbs_reachable = targets
+                .iter()
+                .map(|idx| (*idx, &bbs[*idx]))
+                .filter(|(_, bb)| {
```
Comment (Centril) [inline src/librustc_mir/transform/simplify_try.rs]: @oli-obk I adjusted this check to preserve soundness as discussed. I think the PR should be ready to go now. It's still useful even without `-Z mir-opt-level 2` for `match`es written out and such.

**Comment by Centril** [comment]
Messed up the condition in ^--. All fixed now.

**Comment by Dylan-DPC** [comment]
@oli-obk the linked pr has been merged already. Unblocking this

**Comment by Dylan-DPC** [comment]
@oli-obk sorry wrong pr 😂

File: src/librustc_mir/transform/simplify_try.rs [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+            let (local_tmp_s1, local_0, vf_s1) = match match_set_variant_field(s1) {
+                None => continue,
+                Some(x) => x,
+            };
+            if local_tmp_s0 != local_tmp_s1
+                || vf_s0 != vf_s1
+                || Some((local_0, vf_s0.var_idx)) != match_set_discr(s2)
+            {
+                continue;
+            }
```
Comment (eddyb) [inline src/librustc_mir/transform/simplify_try.rs]: You're not checking that `local_0 != local_1`, which is another soundness condition.
cc @rust-lang/wg-mir-opt

File: src/librustc_mir/transform/simplify_try.rs [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+use rustc_target::abi::VariantIdx;
+use itertools::Itertools as _;
+
+/// Simplifies arms of form `Variant(x) => Variant(x)` to just a move.
+///
+/// This is done by transforming basic blocks where the statements match:
+///
+/// ```rust
+/// _LOCAL_TMP = ((_LOCAL_1 as Variant ).FIELD: TY );
+/// ((_LOCAL_0 as Variant).FIELD: TY) = move _LOCAL_TMP;
```
Comment (eddyb) [inline src/librustc_mir/transform/simplify_try.rs]: This is only sound to collapse if `_LOCAL_TMP` isn't used *anywhere* else (see `promote_consts` for an example of an analysis like that).

But also, there's another issue here... namely, debuginfo.

You need to replace `_LOCAL_TMP` with `((_LOCAL_1 as Variant ).FIELD: TY)` in debuginfo as well, making this a very specialized form of copy-prop, I guess?

cc @rust-lang/wg-mir-opt

File: src/librustc_mir/transform/simplify_try.rs [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+            // Right shape; transform!
+            match &mut s0.kind {
+                StatementKind::Assign(box (place, rvalue)) => {
+                    *place = local_0.into();
+                    *rvalue = Rvalue::Use(Operand::Move(local_1.into()));
+                }
+                _ => unreachable!(),
+            }
+            s1.make_nop();
+            s2.make_nop();
```
Comment (eddyb) [inline src/librustc_mir/transform/simplify_try.rs]: I would replace the third statement and nop the first two. That way, you get the `SourceInfo` of the `Ok(x)` rather than the `x` use (or whatever initialized it).

File: src/librustc_mir/transform/simplify_try.rs (hunk 10 above)
Comment (eddyb) [reply src/librustc_mir/transform/simplify_try.rs]: It would be nice to preserve debuginfo perfectly, but for that we might need a way to merge scopes or something, so that `SourceInfo` from all 3 statements is combined.

Effectively, the resulting `_LOCAL_0 = move _LOCAL_1` needs to be in a superposition of all the scopes.

Which might be hard due to the inlining.

File: src/librustc_mir/transform/simplify_try.rs [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+            projection: &[ProjectionElem::Downcast(_, var_idx), ProjectionElem::Field(field, ty)],
+        } => Some((local, VarField { field, field_ty: ty, var_idx })),
+        _ => None,
+    }
+}
+
+/// Simplifies `SwitchInt(_) -> [targets]`,
+/// where all the `targets` have the same form,
+/// into `goto -> target_first`.
+pub struct SimplifyBranchSame;
```
Comment (eddyb) [inline src/librustc_mir/transform/simplify_try.rs]: I don't think we can do this unless there is no debuginfo attached to scopes used only in the arms, *or* we can convert the debuginfo into a form that switches on the `SwitchInt` operand (plausible in DWARF, not really exposed in LLVM).

Comment (eddyb) [reply src/librustc_mir/transform/simplify_try.rs]: But also this optimization can be rephrased as sinking identical statements from predecessors into the start of a block, followed by `simplify_cfg`.
